src/samVisualize/rendering/forceDirectedGraph.ts

```diff
@@ -0,0 +1,703 @@
+    }
+
+    /**
+     * Updates link paths and node positions every tick.
+     */
+    public ticked(): void {
+        this.links!.selectAll<SVGGElement, LinkDatum>('path').attr('d', this.calculateLinkPath)
+        // // Create a path for the tooltip assistance link, which is just a thicker invisible line
+        // // drawn over existing links (same 'd' value) to make it easier to get a tooltip while hovering over a link.
+        this.nodes!.attr('transform', (d: NodeDatum) => `translate(${d.x}, ${d.y})`)
+        // Once we hit the target after a reheat, begin decaying to long term alpha target
+        if (Math.abs(this.simulation.alpha() - RenderConstants.reheatAlphaTarget) < 0.001) {
+            this.simulation.alphaTarget(RenderConstants.alphaTarget).restart()
+        }
+    }
+
+    /**
+     * Scales down a vector to account for the radius of the target node.
+     * @param dx Difference between source x and target x
+     * @param dy Difference between source y and target y
+     * @param targetNodeRadius Target node radius. Pass radius in live to account for growth animations.
+     * @returns
+     */
+    public scaleLinkVector(
+        dx: number,
+        dy: number,
+        targetNodeRadius: number
+    ): { xComponent: number; yComponent: number } | undefined {
+        const dC = Math.sqrt(dx * dx + dy * dy)
+        if (dC === 0) {
+            return undefined
+        }
+        const d = dC - targetNodeRadius - RenderConstants.arrowheadSize
+        const ratio = d / dC
+        dx *= ratio
+        dy *= ratio
+        return { xComponent: dx, yComponent: dy }
+    }
+
+    // Arrow function because this function is referenced, and it needs to be bound
+    private calculateLinkPath = (d: LinkDatum): string | null => {
+        // LinkDatum source and target properties may be strings or NodeDatums, hence all the type checks.
+        // See d3.SimulationLinkDatum (which LinkDatum extends) documentation:
+
+        //  * For convenience, a link’s source and target properties may be initialized using numeric or string identifiers rather than object references; see link.id.
+        //  * When the link force is initialized (or re-initialized, as when the nodes or links change), any link.source or link.target property which is not an object
+        //  * is replaced by an object reference to the corresponding node with the given identifier.
+        //  * After initialization, the source property represents the source node object.
+
+        // The defined LinkDatum extends d3.SimulationDatum, but overrides the source and target properties to be of type string | NodeDatum,
+        // since in this case links are initialized with string identifiers for source and target properties (resource names).
+        // This ensures that we do not need to check that d.target and d.source are not numbers.
+        const target = d.target
+        const source = d.source
+        if (!_.isString(target) && !_.isString(source)) {
```

Comment:
nit: might help to invert the expression and return early, avoids the wide nested indentations

Reply:
Good point